mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp

```diff
@@ -1479,40 +1475,40 @@ Operation *createCollapsedOp(LinalgType op,
       resultTypes.push_back(newOutput.getType());
   }
 
-  if (isa<linalg::CopyOp>(op)) {
-    return rewriter.create<linalg::CopyOp>(loc, inputOperands[0],
-                                           outputOperands[0]);
-  }
-
-  // Get the iterator types for the operand.
-  SmallVector<utils::IteratorType> iteratorTypes =
-      getCollapsedOpIteratorTypes(op.getIteratorTypesArray(), collapsingInfo);
+  Operation *collapsedOp = clone(
+      rewriter, op, resultTypes,
+      llvm::to_vector(llvm::concat<Value>(inputOperands, outputOperands)));
+
+  // TODO: Find a more general way to determine if op requires explicit
+  // indexing_maps and iterator_types
+  if (isa<linalg::GenericOp>(op)) {
```

Comment:
I dont think you need to do this. All `LinalgOps` have `getIteratorTypesArray` (or should have). Also if you dont clone the op you dont need to set the indexing maps etc. explicitly....

Reply:
Yah i know all `LinalgOp`s have these attrs and the corresponding getters.  But most of the structured ops don't have an explicit attr for indexing maps and iterator types. so if i unconditionally try to set them, i will end up with named ops with additional attrs that aren't at all associated with the implicit ones.  I'm sure there's a way to do this correctly. I just dont know.

I'm not sure how any of this would work without cloning the op (other than having a switch statement checking every single linalg op that exists).  Of course the alternative is to just always convert to a `linalg.generic` op. But I dont like that. I much prefer keeping the named op and that's why I chose to clone.

Reply:
hmm. there is this `kMemoizedIndexingMapsAttrName` that's for named ops.  so i could check if the op has this and then replace it. otherwise replace the expicit indexing maps. but i dont see a similar attr name for iterator types

Reply:
it woudl be nice if there was a general `setIndexingMaps` and `setIteratoryTypes` for arbitrary `LinalgOp`s

Reply:
One alternative is that you could have a interface method to `LinalgOp` that allows you to clone with a given `indexing_maps` and `iterator_types` and just uses the region from the original operation. Each specific op could just implement its own version (including `linalg.generic`).

Reply:
Not a bad idea. Id have to see what it looks like to understand the implications fully but I sense that it might be too demanding to require all ops to implement that.  Would this work for the ops generated from `core_named_ops.py` (can't remember the exact file name right now)?

Reply:
To start with you can just have the method return a `LogicalResult` and the default implementation return `failure()`and only add the implementation for ops that you want to support. Those can be filled in over time.

On second thought maybe instead of going full interface method, just add a templated method in this file for cloning? The default can be for handling named ops, and the generic op can get its specialization

Reply:
Thanks for the feedback. I wont be able to get back to this until this weekend or next, but I'll take a look at your suggestion and keep you updated.